### ISSUE TITLE
Failed to parse expression occurs when constraint has spaces in name

### ISSUE THREAD

**Issue by FreeCAD-Bug-Importer** [FreeCAD/FreeCAD#5750]
Issue imported from https://tracker.freecad.org/view.php?id=3272

* **Reporter:** bgoodr
* **Date submitted:** 11/27/2017
* **FreeCAD version:** 0.17
* **Category:** General
* **Status:** new
* **Tags:** expressions, spreadsheet

# Original report text

Version info:

OS: Ubuntu 17.04
Word size of OS: 64-bit
Word size of FreeCAD: 64-bit
Version: 0.17.12643 (Git)
Build type: None
Branch: master
Hash: 868d9cc6c215ce3a2ab20d454378ec00e1b9ed26
Python version: 2.7.13
Qt version: 4.8.7
Coin version: 4.0.0a
OCC version: 7.1.0
Locale: English/UnitedStates (en_US)

By "external" I mean a spreadsheet that is not a part of the FreeCAD document I am working on. I am choosing to use a separate master spreadsheet because I want that central spreadsheet to control dimensions in constraints used in sketches in multiple FreeCAD documents in the current working directory, such that when I update the values in the "master spreadsheet", the objects in the dependent FreeCAD documents are updated (well, right now they are not updated automatically, and I have to press a button in Assembly2, but that is off-topic from this bug).

After working around the bug in #3271 by manually opening the external spreadsheet file, the fx button does seem to show that it recognizes the syntax and even shows the value, but when I click ok to use that value, I get a "Failed to parse expression" error.  

Use the same attachments as in #3271

The GIF in #3271 also shows this problem.

# Steps to reproduce

1. Extract the tarball.
2. cd into the directory that is extracted so that the current working directory can resolve paths to other files in the set of files.
3. Open up the file: freecad-daily UpsideDownMonitorMount.fcstd
4. Open up the Leg.fcstd file using the button.
5. Open up the MasterSpreadshet.fcstd to work around the bug in #3271 
6. Double click on the sketcher
7. Double click on the constraint with 1.5yd ("Shelf Depth")
8. Click on the fx button.
9. Type in a reference to the alias that is in the spreadsheet: MasterSpreadsheet#Spreadsheet.Sd
10. Notice it says Result: 12.00 "  which is correct: the spreadsheet has 12in in the Sd alias
11. Click Ok
12. Notice the Length field also shows 12.00 " which is correct so far.
13. Click Ok
14. Now see a popup error with window title "Dimensional constraint" and "Failed to parse expression". There is also "Failed to parse expression" in the Report view.

# Other bug information

* **Priority:** normal
* **Severity:** major
* **Category:** General
* **Updated:** 2/6/2021

### Comment by Kunda1 2017-12-09 12:20
Forum thread: https://forum.freecadweb.org/viewtopic.php?f=8&t=25797

**Comment by adrianinsaval** [comment]
external document is unrelated, issue is entirely caused by using a space in the constrain name

**Comment by adrianinsaval** [comment]
#5983 #6593 and #5837 are all related, spaces in constrain name causes many issues

**Comment by luzpaz** [comment]
Ooood, sorry about that. Yea, realthunder expression parser patch hasn't made it through review. Where would you recommend we document this limitation till it gets addressed ?

**Comment by pw-64** [comment]
Unless I have missed it, there should really be a warning or popup expressing that spaces can cause issues.

This is frustrating because I have just finished two day's worth of work and then went round naming my constraints; I then went to change some values in the spreadsheet, only to find my model doesn't update - I eventually traced it back to having spaces in the names.

This now means I, and many others until this is fixed, will have to spend a lot of time going back through every single space-containing thing and re-calculate and re-enter each formula.

I have seen posts own their own forum dating back to 2018 addressing similar issues with spaces causing corruption, so I wonder why such an issue is still not resolved?

**Comment by adrianinsaval** [comment]
Personally I think we should just ban spaces from constraint names (perhaps auto replacing with _). Constraint names should be validated the same as other object/property names.

**Comment by pw-64** [comment]
> Personally I think we should just ban spaces from constraint names (perhaps auto replacing with _). Constraint names should be validated the same as other object/property names.

Since spaces are allowed in body names and the steps inside bodies (I am not familiar with FreeCAD's technical words), it should be made consistent through the whole program. Either allow spaces everywhere and fix whatever is cauing the issue, or disallow spaces everywhere and force underscores. If going with option 2, please add a warning to the user that they entered an invalid name and that the spaces were auto-replaced with the correct separator, underscores.

I saw a comment on the linked forum post talking about the <<name with spaces>> syntax, but I think underscores are a lot more readable since they keep out of the way at the bottom - they aren't so obtrusive. They likened it to the python naming convention where only alphanumeric and underscores are allowed, and must not start with a number.

This would be a much better system than the current mix of unvalidated variable names that can be accepted. I have not tested myself, and it could have changed since their comment, but they also said some built-in workbench(es) already had  behind-the-scenes validation and formatting; if correct, it is strange that within their own workbenches there is inconsistency.

**Comment by pw-64** [comment]
> Ooood, sorry about that. Yea, realthunder expression parser patch hasn't made it through review. Where would you recommend we document this limitation till it gets addressed ?

If it is not possible to create a small popup or a warning text when submitting a space in space-sensitive input, the next best thing is to maybe put a banner on the freecad documentation website saying not to use spaces, and link to this issue. It would also be worth it to create a pinned post on [r/FreeCad](https://www.reddit.com/r/FreeCAD/) as it seems to be quite active and will probably reach a lot more users, especially since this is quite an important, breaking issue.

**Comment by luzpaz** [comment]
@pw-64 can you post to the forum and also bring this up in the discord channel. I've added this to the 1.0 milestone

**Comment by pw-64** [comment]
Which discord channel and forum place should I link to this thread in?
I can also make the reddit post, as I haven't seen one yet.

**Comment by adrianinsaval** [comment]
> Since spaces are allowed in body names and the steps inside bodies (I am not familiar with FreeCAD's technical words), it should be made consistent through the whole program.

I understand why you would not know this since it's a technical difference not obvious to the user, but spaces are not allowed in object names what you are thinking about are called labels. When you hover over an object in the tree view you'll see in the status bar it's actual name which can't be changed after having created the object.

**Comment by pw-64** [comment]
> When you hover over an object in the tree view you'll see in the status bar it's actual name which can't be changed after having created the object.

Ahh, yes I never knew about that. I assumed that internally it either used some kind of UUID or just the display name, since there can only be one. This makes sense though, thank you for informing me.

I still consider this an inconsistency though - why do labels use a hidden & constant name, but sketcher constraints use the actual raw input which can change?

**Comment by GeneGH** [comment]
As I pointed out in the forum, banning spaces is a bit extreme.

I have used spaces for years in sketcher constraint names. Never a problem.

However, if one goes to more complex data links, such as expressions or spreadsheets, there could be problems.

It is a good idea to avoid both spaces and hyphens, but banning is a bit too much.

The best thing would be to either correct the problem functions or warn only on those operations that cannot deal with spaces.

**Comment by pw-64** [comment]
Yes, a permanent ban is extreme. Although, a temporary patch disallowing them while the team waits for the Realthunder patch to be reviewed seems like a good solution to stop any more instances of this issue happening in the mean time.

I encouraged two potential options, one where spaces are allowed, and one with only alphanumeric and underscores. I don't want to push hard for either option.

I think both are valid options. Obviously, option 2 is more restrictive; however, it should ensure there is less possibility of erroneous parsing from current and future invalid characters, such as the possibility of emojis pointed out in the reddit post.

However, I understand spaces are a useful option that people (including myself) do use.

I suppose we could either do a community poll on what kind of labelling & naming convention should be standardised across the whole software, or just go with whichever is currently in progress / easiest to implement.

It is probably a separate issue that would need to be raised, but I feel it would be better to also standardise how these names / labels are actually used / stored (https://github.com/FreeCAD/FreeCAD/issues/5750#issuecomment-1729880473) because this also raises confusion about why some things are referenced by their inputted name vs their "hidden" identifier.

**Comment by adrianinsaval** [comment]
But allowing something that will then be problematic for one of the primary purposes of naming constraints doesn't make much sense. Are spaces really that important? I don't think so, specially if we auto change them to underscores. Is removing spaces really extreme? I don't think so, we already don't allow them for properties and object names. If anything it is more consistent if we don't allow spaces.

**Comment by pw-64** [comment]
That is a valid opinion, although people have expressed that they do value spaces, so it is important we consider how this could impact everyone. This is why I think a poll would be useful, because then the maintainers / contributors can determine whether it would be worthwhile spending time fixing future bugs that could occur with a potentially more breakable system.

I have commented in 3/4 communication posts (the discord isn't very active) for both sides, but I think it should be a community decision by the people that actually use the program, as to what they would like to see.



> auto change them to underscores

If all invalid characters were changed to placeholders in the 'hidden' name, like for bodies, then this would be great. However, since constraints use the raw input as their reference name, auto-changing anything invalid to placeholders could cause confusion and inconsistency; we don't see this behaviour front-end for any of the other names / labels.

**Comment by Roy-043** [comment]
Dashes (minus signs) are also problematic.

Forum topic:
https://forum.freecad.org/viewtopic.php?t=83598

**Comment by MariuszSzafr** [comment]
I dont think its going to get fixed. Spent an hour figuring out why it couldn't reference the clearly named object....there should be a simple error handler(like when you save a file with - / in it) ... like im adding 90 to an angle constraint and you still have to specify deg...its adding to an angle the deg should be known.

**Comment by Yoni528** [comment]
It feels a bit like a waste of the potential of FreeCAD :( 

This is this kind of bug that moves users back to commercial software because "FreeCAD is too buggy for work" and handling it on the UI side would be probably quite an easy task for a developer with QT knowledge

**Comment by hyarion** [comment]
Don't give up just yet :)

**Comment by Yoni528** [comment]
> Don't give up just yet :)

Is there a fix baking in the oven? 😲

**Comment by hyarion** [comment]
No, but I think it is a UX issue that should be handled somehow. Confusion can't be an acceptable solution, though time is a rather limited resources.

I don't see a different between changing constraint names and changing document objects, from the user perspective. Technically, as @adrianinsaval mentioned, document objects has a label that you change and can reference, and it could work the same way for constraint names/labels under the hood. However I don't particularly like that workaround for document objects so maybe we could handle it better for constraints?

As I see it there are two reasons why users might want to name constraints:
* Organization - making it easier to see what's what in sketcher
* Referencing - make use of the name elsewhere

I'm not sure if one is more important than the other. The easiest solution would probably be to just show a warning in the UI when naming the constraint, that the name will not be possible to reference in an expression. At least that could be a first step to limit the confusion of the users.

May I ask how did RT solve it?

**Comment by Yoni528** [comment]
@realthunder are you still on the topic with spaces and dashes in constraint names breaking constraints?

If not do you have recommendations where to start or in what functions to look into? Then I would try to look into it (though I'm not very strong in C++, so no promises :D )

**Comment by aavogt** [comment]
In 1.0.0, I can name a constraint `h` it should be accessible as `Sketch002.Constraints.h` in another sketch. It appears in the completion list when I try to use it. I have to rename it to something longer.

**Comment by adrianinsaval** [comment]
> In 1.0.0, I can name a constraint `h` it should be accessible as `Sketch002.Constraints.h` in another sketch. It appears in the completion list when I try to use it. I have to rename it to something longer.

this is because it clashes with an unit identifier, see https://github.com/FreeCAD/FreeCAD/issues/5770

In the end spaces are no longer accepted in sketch names since #20717 and #22914 will also reject conflicting words, so I guess we can close this issue?

**Comment by yehonatan-regev** [comment]
> > In 1.0.0, I can name a constraint `h` it should be accessible as `Sketch002.Constraints.h` in another sketch. It appears in the completion list when I try to use it. I have to rename it to something longer.
> 
> this is because it clashes with an unit identifier, see [#5770](https://github.com/FreeCAD/FreeCAD/issues/5770)
> 
> In the end spaces are no longer accepted in sketch names since [#20717](https://github.com/FreeCAD/FreeCAD/pull/20717) and [#22914](https://github.com/FreeCAD/FreeCAD/pull/22914) will also reject conflicting words, so I guess we can close this issue?

If spaces are no longer accepted, I guess this issue is obsolete. Thanks for the fix!

**Comment by PaddleStroke** [comment]
@adrianinsaval Ideally it would be best if users could use spaces. You mentioned a RT PR somewhere? Is it overly complex or can we get it to merge?

**Comment by luzpaz** [comment]
Elevate to higher priority ?
